core/src/main/scala/kafka/server/AlterIsrManager.scala

```diff
@@ -0,0 +1,173 @@
+      // Copy current unsent ISRs but don't remove from the map
+      val inflightAlterIsrItems: mutable.Queue[AlterIsrItem] = new mutable.Queue[AlterIsrItem]()
+      unsentIsrUpdates.values().forEach(item => inflightAlterIsrItems.enqueue(item))
+
+      val now = time.milliseconds()
+      lastIsrPropagationMs.set(now)
+
+      buildAndSendRequest(inflightAlterIsrItems.toSeq)
+    }
+  }
+
+  def buildAndSendRequest(inflightAlterIsrItems: Seq[AlterIsrItem]): Unit = {
+    val message = new AlterIsrRequestData()
+      .setBrokerId(brokerId)
+      .setBrokerEpoch(brokerEpochSupplier.apply())
+      .setTopics(new util.ArrayList())
+
+    inflightAlterIsrItems.groupBy(_.topicPartition.topic).foreach(entry => {
+      val topicPart = new AlterIsrRequestData.TopicData()
+        .setName(entry._1)
+        .setPartitions(new util.ArrayList())
+      message.topics().add(topicPart)
+      entry._2.foreach(item => {
+        topicPart.partitions().add(new AlterIsrRequestData.PartitionData()
+          .setPartitionIndex(item.topicPartition.partition)
+          .setLeaderId(item.leaderAndIsr.leader)
+          .setLeaderEpoch(item.leaderAndIsr.leaderEpoch)
+          .setNewIsr(item.leaderAndIsr.isr.map(Integer.valueOf).asJava)
+          .setCurrentIsrVersion(item.leaderAndIsr.zkVersion)
+        )
+      })
+    })
+
+    def responseHandler(response: ClientResponse): Unit = {
+      try {
+        val body: AlterIsrResponse = response.responseBody().asInstanceOf[AlterIsrResponse]
+        handleAlterIsrResponse(body, inflightAlterIsrItems)
+      } finally {
+        // Be sure to clear the in-flight flag to allow future requests
+        if (!inflightRequest.compareAndSet(true, false)) {
+          throw new IllegalStateException("AlterIsr response callback called when no requests were in flight")
+        }
+      }
+    }
+
+    debug(s"Sending AlterIsr to controller $message")
+    controllerChannelManager.sendRequest(new AlterIsrRequest.Builder(message), responseHandler)
+  }
+
+  def handleAlterIsrResponse(alterIsrResponse: AlterIsrResponse, inflightAlterIsrItems: Seq[AlterIsrItem]): Unit = {
+    val data: AlterIsrResponseData = alterIsrResponse.data
+
+    Errors.forCode(data.errorCode) match {
+      case Errors.STALE_BROKER_EPOCH =>
+        warn(s"Broker had a stale broker epoch, retrying.")
+      case Errors.CLUSTER_AUTHORIZATION_FAILED =>
+        warn(s"Broker is not authorized to send AlterIsr to controller")
+        throw Errors.CLUSTER_AUTHORIZATION_FAILED.exception("Broker is not authorized to send AlterIsr to controller")
+      case Errors.NONE =>
+        // Collect partition-level responses to pass to the callbacks
+        val partitionResponses: mutable.Map[TopicPartition, Either[Errors, LeaderAndIsr]] =
+          new mutable.HashMap[TopicPartition, Either[Errors, LeaderAndIsr]]()
+        debug(s"Controller successfully handled AlterIsr request")
```

Comment:
nit: this message would be more useful if we include the response. Perhaps it would be better to log each partition update separately?

Reply:
Should we drop it to trace in that case?